lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+  }
+}
+
+/// Update the stats for a module after a SIL transform has been performed.
+void OptimizerStatsAnalysis::updateModuleStats(TransformationContext &Ctx) {
+  assert(&getModule() == &Ctx.getModule());
+
+  auto &ModStat = getModuleStat();
+  auto OldModStat = ModStat;
+  ModuleStat NewModStat;
+
+  if (OldModStat.FunctionCount == 0) {
+    // This is the first time the stats are computed for the module.
+    // Iterate over all functions in the module and compute the stats.
+    for (auto &F : M) {
+      auto &FuncStat = getFunctionStat(&F);
+      FunctionStat NewFuncStat(&F);
+      processFuncStatHistory(&F, NewFuncStat, Ctx);
+      // Update function stats.
+      FuncStat = NewFuncStat;
+      // Update module stats.
+      NewModStat.addFunctionStat(NewFuncStat);
+    }
+  } else {
+    // Go only over functions that were changed since the last computation.
+    // These are the functions that got invalidated, removed or added.
+    //
+    // If no functions were changed by the last executed transformation, then
+    // the sets of invalidated, deleted and added functions will be empty and
+    // no FunctionStats will be updated.
+    //
+    // FIXME: As a result, the current implementation does not record the fact
+    // about performing a given transformation on a given function, if the
+    // function was not changed during the transformation. This reduced the
+    // amount of recorded information, but makes the history of transformations
+    // on a given function incomplete. If this ever becomes an issue, we can
+    // record all transformations, even if they do not change anything.
+    NewModStat = OldModStat;
+
+    // Process modified functions.
+    while (!InvalidatedFuncs.empty()) {
+      auto *F = InvalidatedFuncs.back();
+      InvalidatedFuncs.pop_back();
+      auto &FuncStat = getFunctionStat(F);
+      auto OldFuncStat = FuncStat;
+      FunctionStat NewFuncStat(F);
+      processFuncStatsChanges(F, OldFuncStat, NewFuncStat, Ctx);
+      NewModStat.subFunctionStat(OldFuncStat);
+      NewModStat.addFunctionStat(NewFuncStat);
+      FuncStat = NewFuncStat;
+    }
+
+    // Process deleted functions.
+    while (!DeletedFuncs.empty()) {
+      auto *F = DeletedFuncs.back();
+      DeletedFuncs.pop_back();
+      auto &FuncStat = getFunctionStat(F);
+      auto OldFuncStat = FuncStat;
+      FunctionStat NewFuncStat;
+      processFuncStatsChanges(F, OldFuncStat, NewFuncStat, Ctx);
+      NewModStat.subFunctionStat(OldFuncStat);
+      Cache->deleteFunctionStat(F);
+    }
+
+    // Process added functions.
+    while (!AddedFuncs.empty()) {
+      auto *F = AddedFuncs.back();
+      AddedFuncs.pop_back();
+      auto &FuncStat = getFunctionStat(F);
+      FunctionStat OldFuncStat;
+      FunctionStat NewFuncStat(F);
+      processFuncStatsChanges(F, OldFuncStat, NewFuncStat, Ctx);
+      NewModStat.addFunctionStat(NewFuncStat);
+      FuncStat = NewFuncStat;
+    }
+  }
+
+  NewModStat.addMemoryStat();
```

Comment:
Why not duplicate this into the if statement and return early. The else statement is pretty long here and you could then get rid of the indentation. I don't have a strong opinion on this though.